### ISSUE TITLE
Self attributes being identified as their parents when being accessed using cls from the classmethod

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#5803]
Trying to solve the issue from https://github.com/microsoft/pyright/issues/5792 on django-types I found another issue:

```python
from typing import ClassVar, List

from typing_extensions import Self


class Foo:
    some_attr: ClassVar[List[Self]]
    attr_in_foo: int

    @classmethod
    def get_some_attr_on_foo(cls):
        reveal_type(cls)
        reveal_type(cls.some_attr)
        for attr in cls.some_attr:
            reveal_type(attr.attr_in_foo)


class Bar(Foo):
    attr_only_in_bar: int

    @classmethod
    def get_some_attr_on_bar(cls):
        reveal_type(cls)

        # This reveals some_attr as List[Self@Foo] and not as List[Self@Bar]
        reveal_type(cls.some_attr)
        for attr in cls.some_attr:
            reveal_type(attr.attr_in_foo)
            # The next line says attr_only_in_bar is Unknown
            reveal_type(attr.attr_only_in_bar)


# The same code here using Bar instead of cls from the classmethod works as expected
reveal_type(Bar.some_attr)
for attr in Bar.some_attr:
    reveal_type(attr.attr_in_foo)
    reveal_type(attr.attr_only_in_bar)
```

The output is:

```
❯ pyright test.py
/tmp/test/test.py
  /tmp/test/test.py:12:21 - information: Type of "cls" is "type[Self@Foo]"
  /tmp/test/test.py:13:21 - information: Type of "cls.some_attr" is "List[Self@Foo]"
  /tmp/test/test.py:15:25 - information: Type of "attr.attr_in_foo" is "int"
  /tmp/test/test.py:23:21 - information: Type of "cls" is "type[Self@Bar]"
  /tmp/test/test.py:26:21 - information: Type of "cls.some_attr" is "List[Self@Foo]"
  /tmp/test/test.py:28:25 - information: Type of "attr.attr_in_foo" is "int"
  /tmp/test/test.py:30:30 - error: Cannot access member "attr_only_in_bar" for type "Foo"
    Member "attr_only_in_bar" is unknown (reportGeneralTypeIssues)
  /tmp/test/test.py:30:25 - information: Type of "attr.attr_only_in_bar" is "Unknown"
  /tmp/test/test.py:34:13 - information: Type of "Bar.some_attr" is "List[Bar]"
  /tmp/test/test.py:36:17 - information: Type of "attr.attr_in_foo" is "int"
  /tmp/test/test.py:37:17 - information: Type of "attr.attr_only_in_bar" is "int"
1 error, 0 warnings, 10 informations
```

This is a MRE from an issue I had when using [this change](https://github.com/sbdchd/django-types/pull/181/files#diff-5fdda2d1a751c2b21dae27794bf7ee0318dd3dd0535819ffc2e1c7b7292d1035R48) to the stubs. e.g. I have a classmethod which will do `for obj in cls.objects.all()` and then access some properties that exists in that obj, but it now thinks `obj` is a `ModelBase` instance and not an instance of that class itself.

**Comment by erictraut** [comment]
I think pyright is correct to generate an error here. Consider the following code, which crashes at runtime.

```python
from typing import ClassVar, Self

class Foo:
    some_attr: ClassVar[list[Self]] = []
    attr_in_foo: int = 0

class Bar(Foo):
    attr_only_in_bar: int = 0

    @classmethod
    def get_some_attr_on_bar(cls):
        for attr in cls.some_attr:
            print(attr.attr_only_in_bar)  # Crash, which is consistent with pyright's error

Foo.some_attr.append(Foo())
Bar.get_some_attr_on_bar()
```

**Comment by bellini666** [comment]
Yes, that indeed crashes, but if the type is there and being set the the metaclass (which is the case with django) it won't.

Also, the crash you have there would happen if doing this as well:

```python
from typing import ClassVar, Self

class Foo:
    some_attr: ClassVar[list[Self]] = []
    attr_in_foo: int = 0

class Bar(Foo):
    attr_only_in_bar: int = 0

    @classmethod
    def get_some_attr_on_bar(cls):
        for attr in cls.some_attr:
            print(attr.attr_only_in_bar)  # Crash, which is consistent with pyright's error

Foo.some_attr.append(Foo())
for attr in Bar.some_attr:
    print(attr.attr_only_in_bar)  # Crash, but pyright doesn't consider this as an error
```

So pyright already allows that, only the classmethod itself is "preventing" that possible runtime crash. Also, in my totally humble opinion, it goes against of what `Self` is supposed to do... I mean, if a situation like that could happen, it would be better for me to type it as `ClassVar[list[Foo]]` to make sure that any callsite using it knows that it is a list of `Foo` or any of its subclasses, and accessing `attr_only_in_bar` would need some kind of narrowing.

But getting to the point to why I opened this issue: The main thing I need to do here is a way to type a class generic attribute, which gets specialized to itself when retrieved. Here are the possibilities:

1) Have a `@property` in the metaclass, which was what the `django-types` was using before but it can't anymore after 1.1.324 due to [it not being compatible with runtime behaviour](https://github.com/microsoft/pyright/issues/5792)

2) Type it with both a `@classmethod` and a `@property` at the same time. Although pyright works correctly with that, it [is not longer supported](https://docs.python.org/3.11/library/functions.html#classmethod) since 3.11, so it would be fair for pyright to also consider it an issue as well at some point.

3) Use the solution I'm trying here, which doesn't fully solve the problem due to this issue.

I'm not sure if there's a fourth option here that I can try, and if there is, I'm totally open to try to use that for that stub.

**Comment by erictraut** [comment]
Yeah, I guess it's always dangerous to use `Self` in a class variable in an invariant context. We may want to make that an error in the future. I think I recall seeing an open bug to that effect in the mypy issue tracker. So that may become an error at some point (necessitating a `# type: ignore`), but for now I will not add an error for this condition.

I've updated pyright's type evaluation logic to do what you've suggested here. I'm still not 100% convinced this is correct. It's a gray area. I decided to make the change to be consistent with mypy's behavior.
